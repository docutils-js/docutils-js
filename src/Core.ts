import {Publisher} from './Publisher';
import * as defaults from './defaults';
import { ConfigSettings, SettingsSpecType } from './types';
import Writer from './Writer';
import Parser from './Parser';
import Reader from './Reader';
import { Settings } from '../gen/Settings';
export { Publisher };

export const defaultUsage = '%%prog [options] [<source> [<destination>]]';
export const defaultDescription = ('Reads from <source> (default is stdin) and writes to <destination> (default is stdout).  See <http://docutils.sf.net/docs/user/config.html> for the full reference.');

/* We need a non command-line parsing based function */

/* We should document all the arguments to this function */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function publish(args: any): void {
    const _defaults = {
        readerName: defaults.defaultReaderName,
        parserName: defaults.defaultParserName,
        usage: defaultUsage,
        description: defaultDescription,
        enableExitStatus: defaults.defaultEnableExitStatus,
    };
    const myArgs = { ..._defaults, ...args };
    const {
        reader, readerName, parser, parserName, writer, writerName,
        settings,
    } = myArgs;
    const pub = new Publisher({
        reader, parser, writer, settings,
    });
    pub.setComponents(readerName, parserName, writerName);
    //pub.publish2({
    //    argv, usage, description, settingsSpec, settingsOverrides, configSection, enableExitStatus,
    //});
}


export interface PublishCmdLineArgs {
reader?: Reader;
readerName?: string;
parser?: Parser;
parserName?: string;
writer?: Writer;
writerName?: string;
settings?: Settings;
settingsSpec?: SettingsSpecType[];
settingsOverrides?: ConfigSettings;
configSection?: string;
enableExitStatus?: boolean;
argv?: string[];
usage?: string,
description?: string;
}

/**
    Set up & run a `Publisher` for command-line-based file I/O (input and
    output file paths taken automatically from the command line).  Return the
    encoded string output also.

    Parameters: see `publish_programmatically` for the remainder.

    - `argv`: Command-line argument list to use instead of ``sys.argv[1:]``.
    - `usage`: Usage string, output if there's a problem parsing the command
      line.
    - `description`: Program description, output for the "--help" option
      (along with command-line option descriptions).
*/
/// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function publishCmdLine(args: PublishCmdLineArgs, cb: any): void {
    const _defaults = {
        readerName: 'standalone',
        parserName: 'restructuredtext',
        usage: defaultUsage,
        description: defaultDescription,
        enableExitStatus: true,
    };
    args = { ..._defaults, ...args };
    const {
        reader, readerName, parser, parserName, writer, writerName,
        settings, settingsSpec, settingsOverrides, configSection,
        enableExitStatus, argv, usage, description,
    } = args;
    const pub = new Publisher({
        reader, parser, writer, settings,
    });
    pub.setComponents(readerName, parserName, writerName);
    pub.publish({
        argv, usage, description, settingsSpec, settingsOverrides, configSection, enableExitStatus,
    }, cb);
}
