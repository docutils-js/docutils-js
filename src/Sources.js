export class Source {
    constructor({ source, sourcePath, encoding }) {
        this.source = source;
        this.sourcePath = sourcePath;
        this.encoding = encoding;
    }
}

export default {
    Source,
};
