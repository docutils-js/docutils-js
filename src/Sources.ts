export default class Source {
    constructor({ source, sourcePath, encoding }) {
        this.source = source;
        this.sourcePath = sourcePath;
        this.encoding = encoding;
    }
}
