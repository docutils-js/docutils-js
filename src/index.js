export const __version__ = '0.14js';
