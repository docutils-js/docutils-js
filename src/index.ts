/* eslint-disable-next-line import/prefer-default-export */
export { parse } from './parse';
export { StringOutput, StringInput } from './io';
export { StandaloneReader } from './readers/standalone';
export { newDocument } from './newDocument';
import * as nodes from './nodes';
import Writer from './Writer';
import Transform from './Transform';
export { Reader } from './Reader';
export { Publisher } from './Publisher'
export { defaults } from '../gen/defaults';
export { pojoTranslate } from './fn/pojoTranslate';
export { htmlTranslate } from './fn/htmlTranslate';
export { RSTParser } from './parsers/restructuredtext';
export { XMLWriter } from './writers/xml'

/* eslint-disable-next-line @typescript-eslint/no-unused-vars,no-unused-vars */
const __docformat__ = 'reStructuredText';

export const __version__ = '0.14js';

export {
    nodes, Writer, Transform,
};
