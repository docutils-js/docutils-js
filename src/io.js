import TransformSpec  from './TransformSpec';
import { createReadStream } from 'fs';
import ErrorOutput from './ErrorOutput';
import Input from './io/Input';
import Output from './io/Output';
import { ApplicationError } from'./Exceptions';

export class StringInput extends Input {
    constructor(...args) {
	super(...args);
	this.sourcePath = '<string>';
    }
	
    read(cb) {
	cb(undefined, this.source);
    }
}

export class StringOutput extends Output {
    constructor(...args) {
	super(...args);
	this.defaultDestinationPath = '<string>'
    }
    
    write(data) {
        //self.destination = self.encode(data) // fixme encoding
	if(Array.isArray(data)) {
	    data = JSON.stringify(data);
	}
	this.destination = data;
	return this.destination;
    }
}

export class FileInput extends Input {
    /* ew, too much logic for a constructor, with side effects etc! */
    constructor(args) {
	super(args);
	let { source, sourcePath, encoding, errorHandler, autoClose,
	      mode } = args;
	if(typeof source === 'undefined' && typeof sourcePath === 'undefined') {
	    throw new ApplicationError("FileInput: Undefined source and sourcePath");
	}
	
	if(autoClose === undefined) {
	    autoClose = true;
	}
	if(mode === undefined) {
	    mode = 'r';
	}
	
	this.autoClose = autoClose;
	this._stderr = new ErrorOutput()
	if(!source) {
	    if(sourcePath) {
		try {
		    this.source = createReadStream(sourcePath, { encoding: 'utf8' });
		}
		catch(error) {
		    console.log(error.stack);
		    throw error;
		}
	    } else {
		this.source = process.stdin;
	    }
	} else {
	    
	    // ??
	}
	if(!sourcePath) {
	    this.sourcePath = this.source.name;
	}
    }

    /* Read and decode a single file and return the data (Unicode string).
     */
    read(cb) {
	setTimeout(() => {
	let data;
	try {
	    /* reading ? */
	    if(this.source === process.stdin) {
		// do stuff
	    } else {
		data = this.source.read();
		if(data === null) {
		    console.log("read returned null ?");
		}
		cb(undefined, data);
	    }
	} catch(error) {
	    console.log(error.stack);
	    cb(error, undefined);
	}
	}, 100);
    }

    readLines() {
	return this.read().splitlines(true);
    }

    close() {
	if(this.source !== process.stdin) {
	    this.source.close();
	}
    }
}

export class FileOutput extends Output {
    
}
