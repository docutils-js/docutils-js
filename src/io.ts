import Input from './io/Input';
import Output from './io/Output';
import { ReadInputCallback, LoggerType } from "./types";

/** Direct string input. */
export class StringInput extends Input {
    public constructor(source: string, logger: LoggerType, sourcePath?: string, encoding?: string, errorHandler?: string) {
        super({source, sourcePath, encoding, errorHandler, logger});
        this.sourcePath = '<string>';
    }

    public read(cb: ReadInputCallback<string|string[]|{}>): void {
        cb(undefined, this.source);
    }
}

export class StringOutput extends Output<string> {
    public constructor(
        logger: LoggerType,
        destination?: string,
        destinationPath?: string,
        encoding?: string,
        errorHandler?: string
    ) {
        super(logger, destination, destinationPath, encoding, errorHandler);
        this.defaultDestinationPath = '<string>';

    }

    public write(data: string): string {
        // self.destination = self.encode(data) // fixme encoding
        if (Array.isArray(data)) {
            data = JSON.stringify(data);
        }
        this.destination = data;
        return this.destination;
    }
}

