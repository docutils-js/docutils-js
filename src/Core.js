import readers from './Readers';
import writers from './Writers';
import { OptionParser } from './FrontEnd';
import { Source } from './Sources';
import { FileInput, FileOutput } from './io';
import Publisher from './Publisher';

export { Publisher };

export const defaultUsage = '%prog [options] [<source> [<destination>]]';
export const defaultDescription = ('Reads from <source> (default is stdin) and writes to <destination> (default is stdout).  See <http://docutils.sf.net/docs/user/config.html> for the full reference.');


export function publishCmdLine(args, cb) {
    const _defaults = {
 readerName: 'standalone',
                        parserName: 'restructuredtext',
                        usage: defaultUsage,
                        description: defaultDescription,
                        enableExitStatus: true,
};
    args = { ..._defaults, ...args };
    const {
 reader, readerName, parser, parserName, writer, writerName,
      settings, settingsSpec, settingsOverrides, configSection,
            enableExitStatus, argv, usage, description,
} = args;
    const pub = new Publisher({
 reader, parser, writer, settings,
});
    pub.setComponents(readerName, parserName, writerName);
    pub.publish({
argv, usage, description, settingsSpec, settingsOverrides, configSection, enableExitStatus,
}, cb);
}
