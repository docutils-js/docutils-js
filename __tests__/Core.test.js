import { Publisher, publishCmdLine, defaultDescription } from '../src/Core'
import { Source } from '../src/Sources';
import { StringInput, StringOutput } from '../src/io'

const defaultArgs = {
    readerName: 'standalone',
    parserName: 'restructuredtext',
    usage: '',
    description: '',
    enableExitStatus: true,
    writerName: 'xml',
};

const defaultSettings = {
    debug: true,
    autoIdPrefix: 'auto',
    idPrefix: '',
};

test.skip('cmdline', () => {
    const description = ('Generates Docutils-native XML from standalone '+
			     'reStructuredText sources.  ' + defaultDescription)
    
    return new Promise((resolve, reject)=>{
	publishCmdLine({ debug: true,
			 debugFn: (msg) => {
			     console.log(`here ${msg}`);
			 },
			 argv: ['in.rst'], writerName: 'xml', description,
		       },
		       (error) => {
			   if(error) {
			       reject(error);
			       return;
			   }
			   resolve();
		       });
    });
});

test('1',  () => {
    const settings = { ...defaultSettings }
    const args = { ...defaultArgs }

    const { readerName, parserName, writerName } = args;
    const source = new StringInput({source: `Block quotes consist of indented body elements:

    This theory, that is mine, is mine.

    -- Anne Elk (Miss)
`});
        const destination = new StringOutput({})
    const pub = new Publisher({source, destination, settings, debug: true, debugFn: console.log});
    pub.setComponents(readerName, parserName, writerName);
    return new Promise((resolve,reject) => {
	pub.publish({}, (error, ...args) => {
	    if(error) {
	    	console.log(error)
		reject(error);
		return;
	    }
	    expect(destination.destination).toMatchSnapshot();
	    resolve();
	});
    });
})

test.each([['Title', "Title\n=====\nParagraph."],
	   ['Random', '* bullet\n* bullet\n\n '],
	   ['Random 2', 'Header 1\n========\nText\n\nHeader 2\n-------'],
	   ['Random 2', 'Test.\nTest2\nTest3\n-----'],
	   ['Random 4', `Test3
-----

This is a test.

* BUllet list 1
* The emacs rst editor is weird.`],
	   ['Emphasis', "*hello*"],
	   ['Emphasis surrounded by text', "stuff *hello* things"],
	   ['Emphasis preceded by text', "stuff *hello*"],
	   ['Emphasis followed by text', "*hello* test"],
	   ['Strong', '**hello**'],
	   ['Emphasis and inline', '*hello* and **goodbye**'],
	   ['Inline followed by emphasis', '**hello** and *goodbye*'],
	  ] )("%s", (a,raw) => {
	      const settings = { ...defaultSettings }
	      const args = { ...defaultArgs }
	      
	      const { readerName, parserName, writerName } = args;
//	      console.log(raw);
	      const source = new StringInput({source: raw});
	      const destination = new StringOutput({})
	      const pub = new Publisher({source, destination, settings, debug: true, debugFn: console.log});
	      pub.setComponents(readerName, parserName, writerName);
	      return new Promise((resolve, reject) => {
		  /* {argv, usage, description, settingsSpec, settingsOverrides, configSection, enableExitStatus }*/
		  pub.publish({}, (error, ...args) => {
		      if(error) {
			  reject(error);
			  return;
		      }
		      expect(destination.destination).toMatchSnapshot();
//		      console.log(destination.destination);
		      resolve();
		  });
	      });
	  });
						   

