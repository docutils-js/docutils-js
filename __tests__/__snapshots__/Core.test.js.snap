// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1 1`] = `
"<!-- generated by Docutils 0.14js -->
<document source=\\"True\\">
<paragraph>Block quotes consist of indented body elements:</paragraph><block_quote><paragraph>This theory, that is mine, is mine.</paragraph><attribution>Anne Elk (Miss)</attribution></block_quote></document>"
`;

exports[`Emphasis 1`] = `
"<!-- generated by Docutils 0.14js -->
<document source=\\"True\\">
<paragraph><emphasis>hello</emphasis></paragraph></document>"
`;

exports[`Emphasis and inline 1`] = `
"<!-- generated by Docutils 0.14js -->
<document source=\\"True\\">
<paragraph><emphasis>hello</emphasis> and <strong>goodbye</strong></paragraph></document>"
`;

exports[`Emphasis followed by text 1`] = `
"<!-- generated by Docutils 0.14js -->
<document source=\\"True\\">
<paragraph><emphasis>hello</emphasis> test</paragraph></document>"
`;

exports[`Emphasis preceded by text 1`] = `
"<!-- generated by Docutils 0.14js -->
<document source=\\"True\\">
<paragraph>stuff <emphasis>hello</emphasis></paragraph></document>"
`;

exports[`Emphasis surrounded by text 1`] = `
"<!-- generated by Docutils 0.14js -->
<document source=\\"True\\">
<paragraph>stuff <emphasis>hello</emphasis> things</paragraph></document>"
`;

exports[`Inline followed by emphasis 1`] = `
"<!-- generated by Docutils 0.14js -->
<document source=\\"True\\">
<paragraph><strong>hello</strong> and <emphasis>goodbye</emphasis></paragraph></document>"
`;

exports[`Random 1`] = `
"<!-- generated by Docutils 0.14js -->
<document source=\\"True\\">
<bullet_list bullet=\\"*\\">
<list_item>
<paragraph>bullet</paragraph></list_item></bullet_list><bullet_list bullet=\\"*\\"><list_item><paragraph>bullet</paragraph></list_item></bullet_list></document>"
`;

exports[`Random 2 1`] = `
"<!-- generated by Docutils 0.14js -->
<document source=\\"True\\">
<section ids=\\"Header 1\\" names=\\"Header 1\\">
<title>Header 1</title><paragraph>Text</paragraph><section ids=\\"Header 2\\" names=\\"Header 2\\"><title>Header 2</title><system_message line=\\"6\\"><paragraph>Title underline too short.</paragraph><literal_block>Header 2
-------</literal_block></system_message></section></section></document>"
`;

exports[`Random 2 2`] = `
"<!-- generated by Docutils 0.14js -->
<document source=\\"True\\">
<paragraph>Test.
Test2
Test3
-----</paragraph></document>"
`;

exports[`Random 4 1`] = `
"<!-- generated by Docutils 0.14js -->
<document source=\\"True\\">
<section ids=\\"Test3\\" names=\\"Test3\\">
<title>Test3</title><paragraph>This is a test.</paragraph><bullet_list bullet=\\"*\\"><list_item><paragraph>BUllet list 1</paragraph></list_item></bullet_list><bullet_list bullet=\\"*\\"><list_item><paragraph>The emacs rst editor is weird.</paragraph></list_item></bullet_list></section></document>"
`;

exports[`Strong 1`] = `
"<!-- generated by Docutils 0.14js -->
<document source=\\"True\\">
<paragraph><strong>hello</strong></paragraph></document>"
`;

exports[`Title 1`] = `
"<!-- generated by Docutils 0.14js -->
<document source=\\"True\\">
<section ids=\\"Title\\" names=\\"Title\\">
<title>Title</title><paragraph>Paragraph.</paragraph></section></document>"
`;
